tools/autograd/derivatives.yaml

```diff
@@ -1598,7 +1598,7 @@
   self: soft_margin_loss_backward(grad, self, target, reduction)
 
 - name: relu(Tensor self) -> Tensor
-  self: threshold_backward(grad, self, 0)
+  self: threshold_backward(grad, result, 0)
```

Comment:
You can remove the inplace version below for both.
The non-inplace version will be used for both automatically.

Reply:
When removing the inplace versions, test_leaky_relu_inplace_with_neg_slope_... result in an AssertionError "RuntimeError not raised"

Reply:
It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values, if it is doable with PyTorch

Reply:
It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu_, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values - however it is done in PyTorch

Reply:
It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu_, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values:
```yaml
self: "negative_slope >= 0 ? leaky_relu_backward(grad, result, negative_slope, false) : leaky_relu_backward(grad, self, negative_slope, false)"
```

Reply:
It seems that c10::Scalar does not allow >=, thus we may have to add those to Scalar.cpp - how should I test if it actualy does what it should?

Reply:
Ho sorry, I forgot leaky relu doesn't have the same formula for inplace. My bad.
You can remove only relu_ then, thanks!

Reply:
It should not work for leaky_relu as well when we replace "self" with "return".

Using "return" is only valid as a replacement for "self" when negative_slope value is positive, because the output has to behave the same regarding the 0 threshold as the input - with a negative_slope < 0 the output will be positive for self > 0 and self < 0, so that "result" becomes indistinguishable regarding the threshold, which would result in wrong gradients.

Thus, using the less VRAM needy version for leaky_relu will make its behavior equivalent to reaky_relu_, so that it should throw a Not Supported Error for negative_slope < 0 as well. However, this would change its behavior, which may result in errors for users. I recommend to either not improve the VRAM need for leaky_relu, which is up to 20% savings for a Conv2d net I use, or we should distinguish regarding the negative_slope and use "self" for negative negative_slopes

Reply:
Should we remove the error message from Activation.cpp (line 116)?

Reply:
Yes, you can add back the inplace leaky relu version.

Reply:
@albanD It seems that both "result" and "self" will be stored for LeakyReLU for the above solution, which are set at compile time. How may someone store backpropagation variables at run time, or how should we handle a negative_slope < 1 other than using the old leaky_relu or not allowing a negative_slope for the function?

Reply:
If we add back the inplace leaky relu version, we will need to use "self" for the other relu version

Reply:
If we add back the inplace leaky relu version, we will need to use "self" for the leaky relu

Reply:
Sorry for the delay.
From checking the code, I think we want:
- Keep the current change to relu of always replacing self by result. It means that we can remove the inplace entry as the inplace and out of place now have the same formula.
- Leave leaky relu as it is today in master.
  - Using result in the out of place version would mean that we cannot  compute the gradient for negative slopes.
  - We could add a system to choose what is going to be saved based on the value of negative_slope. But that would be another feature, outside of the scope of the current PR.